RFC: `bee-connect-tcp` crate (#1)
iotaledger/bee-rfcs: text/0000-bee-connect-tcp-crate.md

+ Feature Name: `bee-connect-tcp-crate`
+ Start Date: 2019-09-05

> + RFC PR: [iotaledger/bee-rfcs#0000](https://github.com/iotaledger/bee-rfcs/pull/0000)

Reviewer: Change accordingly.

> + Bee issue: [iotaledger/bee#0000](https://github.com/iotaledger/bee/issues/0000)

Reviewer: Change accordingly.

# Summary

> Build `bee-connect-tcp`, a crate responsible for byte transmission from and to neighbored bees via TCP.

Reviewer: ```suggestion
Specifications for `bee-connect-tcp`, a crate responsible for byte transmission from and to paired nodes via TCP.
```

# Motivation

> The IOTA Tangle is a distributed ledger. That requires multiple bee nodes to form a network and exchange gossip. To

Reviewer: ```suggestion
The IOTA Tangle is a distributed ledger that requires multiple nodes to form a network and exchange gossip. To
```

> build such a gossip protocol, one must allow bee nodes to exchange information by sending/receiving bytes to/from

Reviewer: ```suggestion
build such a gossip protocol, one must allow nodes to exchange information by sending/receiving bytes to/from
```

> neighbored bees.

Reviewer: ```suggestion
other paired nodes.
```

# Detailed design

## General

This crate does not consider any IOTA concepts such as transactions. Instead, it handles the transmission of bytes
without knowing the underlying semantics of these bytes (whether they represent an IOTA transaction or a transaction request).
This crate works closely together with `bee-gossip`, which is responsible for translating between IOTA concepts and their
pure byte representations as well as handling incoming and outgoing traffic.

Further, every `bee-connect` implementation uses a specific protocol (e.g. TCP). Multiple `bee-connect` crate implementations
with different protocols can be installed simultaneously. EEE makes this possible. Each `bee-connect` crate simply ignores
actions related to other protocols.

## Exposed Interface

### NodeID

`NodeID` instances describe how to reach a single external node. They specify the protocol and address of such a node so
that a connection can be established based on those information. They must include a protocol identifier as well as an
unique address.

> This struct should actually be independent of the `bee-connect-tcp` crate since it covers every `bee-connect crate`.

Reviewer: ```suggestion
This struct should actually be independent of the `bee-connect-tcp` crate since it covers every `bee-connect` crate.
```
